src/Concordium/Client/GRPC2.hs

```diff
@@ -2552,20 +2852,234 @@ getAccountInfoV2 ::
 
+-- |Send a request to a unary V2 GRPC API endpoint and return the result.
+withUnaryV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ MethodOutput CS.Queries m ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |A mapping of the result.
+    (GRPCResult o -> b) ->
+    ClientMonad n b
+withUnaryV2 method input k = withGRPCCoreV2 callHelper k
+    where
+        -- This is here so we may leverage `withGRPCCoreV2`.
+        callHelper client = do
+            res <- rawUnary method client input
+            return $ fmap RawUnaryOutput res
+
+-- |Invoke a streaming V2 GRPC API endpoint and return the (collected) results in a sequence.
+withServerStreamCollectV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ Seq (MethodOutput CS.Queries m) ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |A mapping of the collected result.
+    (GRPCResult o -> b) ->
+    ClientMonad n b
+withServerStreamCollectV2 method input =
+    withServerStreamCallbackV2 method input mempty handler
+    where handler acc o = return $ acc <> pure o
+
+-- |Invoke a streaming GRPC API endpoint and return the (collected) results.
+-- Takes a `fold`-like callback and an accumulator, and returns the result
+-- of folding through each stream element, once the stream terminates.
+withServerStreamCallbackV2 ::
+    ( HasMethod CS.Queries m,
+      MonadIO n,
+      i ~ MethodInput CS.Queries m,
+      o ~ MethodOutput CS.Queries m ) =>
+    -- |The procedure to invoke.
+    RPC CS.Queries m ->
+    -- |The procedure input.
+    i ->
+    -- |An initial `fold`-like accumulator that
+    -- is updated with the handler each time a
+    -- stream element arrives.
+    a ->
+    -- |A `fold`-like handler which is used
+    -- to process a stream object and update
+    -- the accumulator.
+    (a -> o -> ClientIO a) ->
+    -- |A mapping of the accumulated result.
+    (GRPCResult a -> b) ->
+    ClientMonad n b
+withServerStreamCallbackV2 method input acc handler k =
+    withGRPCCoreV2 callHelper k
+    where
+        -- This is simply a handler which conforms to the one required by
+        -- `rawStreamServer`. This is here, so we may ignore the response
+        -- headers in the handler, just for the sake of ergonomics.
+        handler' acc' _hds streamObj = handler acc' streamObj
+        -- Helper that invokes the streaming call with the above callback.
+        -- This is here so we may leverage `withGRPCCoreV2`.
+        callHelper client = do
+            res <- rawStreamServer method client acc input handler'
+            return $ fmap ServerStreamOutput res
+
+-- |Setup a GRPC client and invoke a request helper which uses the client.
+-- The output is interpreted using the function given in the second parameter.
+withGRPCCoreV2 ::
```

Comment:
I am not a fan of copying this. THis function is subtle. Can we not just reuse the one we already have for V1?

Reply:
They are identical up to a few changes. Specifically it uses `outputGRPCV2` rather than `outputGRPC`, and a helper rather than `rawUnary`. It's technicallt copied sincr I commented out the old version in the commit. It is not possible to use the V1 version for the streaming calls though. What do you propose?

Reply:
So the way the client is used now it is good enough I think. The only time we would use "getBlocks/getFinalizedBlocks" is for tailing transactions, which means nothing else is happening then anyhow.

Also the way it is used in the wallet-proxy it is fine. So we will probably have to just live with this in order to get this done. It is unfortunate the Haskell library is shitty. I don't want to spend much time revising this further unless it becomes an issue.

So for start we make it clear that there is this limitation and move on.

Reply:
Did this answer your question?

Reply:
I believe so 👍🏼 so do I understand this correctly: The issue comes from the client instance currently working like a singleton with a mutex?

So this is okay for the _non-streaming_ and _non-tailing, streaming_ calls as their duration for acquiring the mutex is likely to be minute.

For tailing calls, the mutex is acquired for the duration of the call, but I suppose this is OK since these are not used in the wallet-proxy.

So this limitation should be stated in the documentation (perhaps) of `withGRPCCoreV2` and perhaps `getBlocksV2`/`getFinalizedBlocksV2`, for future reference.

Reply:
Do we need to set a request timeout for the tailing calls?

Reply:
It's not a mutex in normal operation. It is reader-writer lock protected.  So multiple queries are supported in paralell, but if things go wrong then we need to re-establish the connection and that requires a writer lock.

The issue is that if there is a long-running (and non-productive) streaming call then it will have the reader lock acquired for the full duration of the stream. Which will prevent connection-reset since the writer lock cannot be acquired.

It should not be an issue in practice the way the client is used, we just need to be aware of it.

Reply:
> Do we need to handle request timeout differently for the tailing calls?

Yeah.

But let's not go down that rabit hole now. Let's do this as a last thing after we have a working client. We don't need the getFinalizedBlocks and getBlocks calls to get a working client mirroring what it does now. We can continue polling as we do now for transaction statuses.

Reply:
> But let's not go down that rabit hole now.

Right, just as to make note of this.